src/middleware/r2Middleware.ts

```diff
@@ -0,0 +1,150 @@
+import { CACHE_HEADERS } from '../constants/cache';
+import { Context } from '../context';
+import { R2Provider } from '../providers/r2Provider';
+import responses from '../responses';
+import { hasTrailingSlash, isDirectoryPath } from '../utils/path';
+import type { Middleware } from './middleware';
+import { LATEST_RELEASES } from '../constants/r2Prefixes';
+import { Request } from '../routes/request';
+import { renderDirectoryListing } from '../utils/directoryListing';
+import { parseConditionalHeaders } from '../utils/request';
+import { once } from '../utils/memo';
+
+const getProvider = once((ctx: Context) => new R2Provider({ ctx }));
+
+export class R2Middleware implements Middleware {
+  async handle(request: Request, ctx: Context): Promise<Response> {
+    const path = getR2Path(request);
+    const isPathADirectory = isDirectoryPath(path);
+
+    return isPathADirectory
+      ? handleDirectory(request, path, ctx)
+      : handleFile(request, path, ctx);
+  }
+}
+
+async function handleDirectory(
+  request: Request,
+  r2Path: string,
+  ctx: Context
+): Promise<Response> {
+  if (!hasTrailingSlash(request.urlObj.pathname)) {
+    const url = request.unsubtitutedUrl ?? request.urlObj;
+    return Response.redirect(url + '/', 301);
+  }
+
+  const result = await getProvider(ctx).readDirectory(r2Path, {
+    // /docs lists the nodejs/release directory - don't want to include the
+    //  files in there for that path
+    listFiles: !request.urlObj.pathname.startsWith('/docs'),
+  });
+  if (result === undefined) {
```

Comment:
In general, if statements and constant assignments have empty lines between each other, I believe it's just preference, but I think it makes it easier to read, at least for me. But I'm fine with the current code.